cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/partitioned/DynamicPartitioningOutputCommitter.java

```diff
@@ -218,53 +248,57 @@ public void abortJob(JobContext context, JobStatus.State state) throws IOExcepti
       } finally {
         // clear this so that we only attempt rollback once in case it gets called multiple times
         outputDataset = null;
-        relativePaths = null;
+        partitionsToAdd = null;
       }
     }
     super.abortJob(context, state);
   }
 
-// Copied from superclass to enable usage of it, because our 'from' and 'to' locations are different.
+  // Based off method from FileOutputCommitter: mergePaths(FileSystem fs, final FileStatus from, final Path to);
+  // It has been modified to handle the case of two DynamicPartitioner MR creating (while allowing append) the output
+  // directory for a Partition. If both find that the partition directory does not exist at the same time, then the
+  // second one that does the rename from the 'from' directory to the 'to' directory would have actually resulted
+  // in moving the 'from' directory INTO the 'to' directory, since it was created by the first MR.
   /**
    * Merge two paths together.  Anything in from will be moved into to, if there
    * are any name conflicts while merging the files or directories in from win.
-   * @param fs the File System to use
+   * @param fc the FileContext to use
    * @param from the path data is coming from.
    * @param to the path data is going to.
    * @throws IOException on any error
    */
-  private void mergePaths(FileSystem fs, final FileStatus from, final Path to) throws IOException {
+  private void mergePaths(FileContext fc, final FileStatus from, final Path to) throws IOException {
     if (from.isFile()) {
-      if (fs.exists(to)) {
-        if (!fs.delete(to, true)) {
+      if (fc.util().exists(to)) {
+        if (!fc.delete(to, true)) {
           throw new IOException("Failed to delete " + to);
         }
       }
 
-      if (!fs.rename(from.getPath(), to)) {
-        throw new IOException("Failed to rename " + from + " to " + to);
-      }
+      fc.rename(from.getPath(), to);
     } else if (from.isDirectory()) {
-      if (fs.exists(to)) {
-        FileStatus toStat = fs.getFileStatus(to);
-        if (!toStat.isDirectory()) {
-          if (!fs.delete(to, true)) {
-            throw new IOException("Failed to delete " + to);
-          }
-          if (!fs.rename(from.getPath(), to)) {
-            throw new IOException("Failed to rename " + from + " to " + to);
-          }
-        } else {
-          //It is a directory so merge everything in the directories
-          for (FileStatus subFrom: fs.listStatus(from.getPath())) {
-            Path subTo = new Path(to, subFrom.getPath().getName());
-            mergePaths(fs, subFrom, subTo);
-          }
+      if (!fc.util().exists(to)) {
+        //it does not exist just rename
+        try {
+          fc.rename(from.getPath(), to);
+          return;
+        } catch (FileAlreadyExistsException e) {
+          // This race condition can happen if two MR (DynamicPartitioner) are appending to partitions, and both
```

Comment:
It seems cleaner to expect that the FileAlreadyExistsException in the case of that described race condition, for which I commented there when it can happen.